http-client/src/main/java/io/airlift/http/client/spnego/SpnegoAuthentication.java

```diff
@@ -193,6 +187,62 @@ public boolean matches(String type, URI uri, String realm)
         return NEGOTIATE.equalsIgnoreCase(type);
     }
 
+    private LoginContext loadLoginContext()
+    {
+        try {
+            LoginContext loginContext = new LoginContext("", null, null, new Configuration()
+            {
+                @Override
+                public AppConfigurationEntry[] getAppConfigurationEntry(String name)
+                {
+                    ImmutableMap.Builder<String, String> optionsBuilder = ImmutableMap.builder();
+                    optionsBuilder.put("refreshKrb5Config", "true");
+                    optionsBuilder.put("doNotPrompt", "true");
+                    optionsBuilder.put("useKeyTab", "true");
+                    if (LOG.isDebugEnabled()) {
+                        optionsBuilder.put("debug", "true");
+                    }
+
+                    if (keytab != null) {
+                        optionsBuilder.put("keytab", keytab.getAbsolutePath());
+                    }
+
+                    if (credentialCache != null) {
+                        optionsBuilder.put("ticketCache", credentialCache.getAbsolutePath());
+                        optionsBuilder.put("useTicketCache", "true");
+                        optionsBuilder.put("renewTGT", "true");
+                    }
+
+                    if (principal != null) {
+                        optionsBuilder.put("principal", principal);
+                    }
+
+                    return new AppConfigurationEntry[] {
+                            new AppConfigurationEntry(Krb5LoginModule.class.getName(), REQUIRED, optionsBuilder.build())
+                    };
+                }
+            });
+
+            loginContext.login();
+            return loginContext;
+        }
+        catch (LoginException e) {
+            throw Throwables.propagate(e);
+        }
+    }
+
+    private LoginContext getLoginContext()
+    {
+        try {
+            // this is the only entry in the cache
+            return loginContextCache.get("loginContext");
+        }
+        catch (ExecutionException e) {
+            Throwables.propagate(e);
```

Comment:
This should get the cause from `ExecutionException`, since that is always a wrapper exception